File: IndividualAssignment1/Code/Python/Benchmark.py


import time
import statistics
import csv
from Matrix import make_random_matrix, multiply_naive, multiply_blocked, multiply_numpy

IMPLEMENTATIONS = {
    "naive": multiply_naive,
    "blocked": multiply_blocked,
    "numpy": multiply_numpy,
}

SIZES = [64, 128, 256, 512]
REPEATS = 5
BLOCK_SIZE = 32
CSV_FILE = "/Users/rafasuarzz/Documents/IndividualAssignment/Results/results_python.csv"  # poner None si no quieres CSV
# ----------------------------------------------------------------------


def run_once(impl_name, n, block):
    A = make_random_matrix(n, seed=123)
    B = make_random_matrix(n, seed=456)
    impl = IMPLEMENTATIONS[impl_name]
    start = time.perf_counter()
    if impl_name == "blocked":
        impl(A, B, block)
    else:
        impl(A, B)
    end = time.perf_counter()
    return end - start


def benchmark(impl_name, n, repeats, block=32):
    times = []
    for r in range(repeats):
        t = run_once(impl_name, n, block)
        times.append(t)
    return times


def pretty_print(name, n, times):
    """Imprime y devuelve estadísticas resumidas."""
    mean = statistics.mean(times)
    std = statistics.stdev(times) if len(times) > 1 else 0.0
    print(f"Impl={name}, n={n}, runs={len(times)}")
    print(f"  min={min(times):.6f}s mean={mean:.6f}s std={std:.6f}s max={max(times):.6f}s\n")
    return mean, std, min(times), max(times)


def main():
    results = []
    print("\n=== Benchmarking all implementations ===\n")

    for impl in IMPLEMENTATIONS.keys():
        print(f"--- {impl.upper()} IMPLEMENTATION ---")
        for n in SIZES:
            times = benchmark(impl, n, REPEATS, BLOCK_SIZE)
            mean, std, tmin, tmax = pretty_print(impl, n, times)
            results.append({
                "impl": impl,
                "n": n,
                "mean": mean,
                "std": std,
                "min": tmin,
                "max": tmax
            })
        print("-" * 60)

    if CSV_FILE:
        with open(CSV_FILE, "w", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=["impl", "n", "mean", "std", "min", "max"])
            writer.writeheader()
            writer.writerows(results)
        print(f"Resultados guardados en {CSV_FILE}\n")


if __name__ == "__main__":
    main()

File: IndividualAssignment1/Code/C++/Benchmark.cpp
#include "Matrix.h"
#include <chrono>
#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>
#include <fstream>
#include <algorithm>
#include <string>

int main() {
    // -----------------------------------------------------
    // Configuración general (puedes modificar aquí)
    // -----------------------------------------------------
    std::vector<int> sizes = {64, 128, 256, 512}; // tamaños de las matrices
    int repeats = 5;                              // repeticiones por tamaño
    int block = 32;                               // tamaño de bloque
    std::vector<std::string> impls = {"naive", "blocked"};
    std::string csv_file = "/Users/rafasuarzz/Documents/IndividualAssignment/Results/results_cpp.csv";     // poner "" si no quieres CSV
    // -----------------------------------------------------

    std::cout << "\n=== Benchmarking all C++ implementations ===\n" << std::endl;

    std::ofstream csv;
    if (!csv_file.empty()) {
        csv.open(csv_file);
        csv << "impl,n,mean,std,min,max\n";
    }

    for (const auto& impl : impls) {
        std::cout << "--- " << impl << " IMPLEMENTATION ---" << std::endl;

        for (int n : sizes) {
            std::vector<double> times;
            times.reserve(repeats);

            for (int r = 0; r < repeats; r++) {
                auto A = random_matrix(n, 1000 + r);
                auto B = random_matrix(n, 2000 + r);

                auto t0 = std::chrono::high_resolution_clock::now();
                if (impl == "blocked") {
                    blocked_multiply(A, B, block);
                } else {
                    naive_multiply(A, B);
                }
                auto t1 = std::chrono::high_resolution_clock::now();
                std::chrono::duration<double> diff = t1 - t0;
                times.push_back(diff.count());
            }

            double sum = 0;
            for (double v : times) sum += v;
            double mean = sum / times.size();
            double variance = 0;
            for (double v : times) variance += (v - mean) * (v - mean);
            variance /= times.size();
            double stddev = std::sqrt(variance);
            double min_time = *std::min_element(times.begin(), times.end());
            double max_time = *std::max_element(times.begin(), times.end());

            std::cout << "Impl=" << impl
                      << ", n=" << n
                      << ", runs=" << repeats << std::endl;
            std::cout << "  min=" << std::fixed << std::setprecision(6) << min_time
                      << "s mean=" << mean
                      << "s std=" << stddev
                      << "s max=" << max_time << "s\n" << std::endl;

            if (csv.is_open()) {
                csv << impl << "," << n << ","
                    << mean << "," << stddev << ","
                    << min_time << "," << max_time << "\n";
            }
        }
        std::cout << "----------------------------------------------------\n" << std::endl;
    }

    if (csv.is_open()) {
        csv.close();
        std::cout << "Resultados guardados en " << csv_file << std::endl;
    }

    return 0;
}
